File: src/main/java/cat/inspiracio/calculator/Sphere.java
/*	Copyright 2011 Alexander Bunkenburg alex@inspiracio.cat
 * 
 * This file is part of Complex Calculator.
 * 
 * Complex Calculator is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Complex Calculator is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Complex Calculator. If not, see <http://www.gnu.org/licenses/>.
 * */
package cat.inspiracio.calculator;

import cat.inspiracio.complex.Complex;
import cat.inspiracio.complex.package$;
import cat.inspiracio.numbers.ECList;

import java.awt.*;

// Referenced classes of package bunkenba.calculator:
//            WorldRepresentation, DoubleBuffer, Drawing, Matrix44, 
//            Vector3, World

final class Sphere extends WorldRepresentation{

    private static int MARKLENGTH = 2;

	//State -------------------------------------------------------
	
    private Matrix44 R;
    private Matrix44 R1;
    private double xyscale;

    private final Complex marks[] = {
                Real(0),
                infinity,
                Real(1),
                Cartesian(-1D, 0.0D),
                Cartesian(0, 1),
                Cartesian(0.0D, -1D)
    };

	//Constructor -------------------------------------------------
	
    Sphere(World world){
        super(world);
        R = new Matrix44();
        R1 = new Matrix44();
        reset();
    }
    
    //Methods ------------------------------------------------------

    @Override void drawComplex(Drawing drawing, Complex z){
        double d;
        double d1;
        double d2;
        if( z.isFinite() ){
            double a = abs(z);
            double d3 = sqr( a );
            double d5 = 1.0D + d3;
            d = Re(z) / d5;
            d1 = d3 / d5 - 0.5D;
            d2 = Im(z) / d5;
        } else{
            d = 0.0D;
            d1 = 0.5D;
            d2 = 0.0D;
        }
        double d4 = R.data[2][0] * d + R.data[2][1] * d1 + R.data[2][2] * d2 + R.data[2][3];
        if(d4 <= 0.0D){
            double d6 = R.data[0][0] * d + R.data[0][1] * d1 + R.data[0][2] * d2 + R.data[0][3];
            double d7 = R.data[1][0] * d + R.data[1][1] * d1 + R.data[1][2] * d2 + R.data[1][3];
            drawing.cross((int)(d6 * xyscale + (double)getSize().width * 0.5D), (int)(-d7 * xyscale + (double)getSize().height * 0.5D), MARKLENGTH);
            drawing.move(2, -2);
            drawing.drawString(z.toString());
        }
    }

    void drawECList(Drawing drawing, ECList eclist){
        Point point = new Point();
        if(eclist != null){
            boolean flag = isFrontC(eclist.head(), point);
            drawing.moveTo(point.x, point.y);
            for(; eclist != null; eclist = eclist.tail())
                if(isFrontC(eclist.head(), point)){
                    if(flag)
                        drawing.lineTo(point.x, point.y);
                    else
                        drawing.moveTo(point.x, point.y);
                    flag = true;
                } else{
                    flag = false;
                }
        }
    }

    private Complex f3dC(Vector3 v){
        if(v.y == 0.5D)
            return infinity;
        if(v.y == -0.5D)
            return Real(0);
        else
            return Cartesian(v.x / (0.5D - v.y), v.z / (0.5D - v.y));
    }

    private boolean isFrontC(Complex z, Point point){
        double d;
        double d1;
        double d2;
        if( z.isFinite() ){
            double d3 = sqr( abs(z) );
            double d5 = 1.0D + d3;
            d = Re(z) / d5;
            d1 = d3 / d5 - 0.5D;
            d2 = Im(z) / d5;
        } else{
            d = 0.0D;
            d1 = 0.5D;
            d2 = 0.0D;
        }
        double d4 = R.data[2][0] * d + R.data[2][1] * d1 + R.data[2][2] * d2 + R.data[2][3];
        if(d4 <= 0.0D){
            double d6 = R.data[0][0] * d + R.data[0][1] * d1 + R.data[0][2] * d2 + R.data[0][3];
            double d7 = R.data[1][0] * d + R.data[1][1] * d1 + R.data[1][2] * d2 + R.data[1][3];
            point.x = (int)(d6 * xyscale + (double)getSize().width * 0.5D);
            point.y = (int)(-d7 * xyscale + (double)getSize().height * 0.5D);
            return true;
        } else{
            return false;
        }
    }

    @Override void lineTo(Drawing drawing, Complex z){
        double d;
        double d1;
        double d2;
        if( z.isFinite() ){
            double d3 = sqr( abs(z) );
            double d5 = 1.0D + d3;
            d = Re(z) / d5;
            d1 = d3 / d5 - 0.5D;
            d2 = Im(z) / d5;
        } else{
            d = 0.0D;
            d1 = 0.5D;
            d2 = 0.0D;
        }
        double d4 = R.data[2][0] * d + R.data[2][1] * d1 + R.data[2][2] * d2 + R.data[2][3];
        if(d4 <= 0.0D){
            double d6 = R.data[0][0] * d + R.data[0][1] * d1 + R.data[0][2] * d2 + R.data[0][3];
            double d7 = R.data[1][0] * d + R.data[1][1] * d1 + R.data[1][2] * d2 + R.data[1][3];
            drawing.lineTo(
                    (int)(d6 * xyscale + (double)getSize().width * 0.5D),
                    (int)(-d7 * xyscale + (double)getSize().height * 0.5D)
            );
        }
    }

    @Override void moveTo(Drawing drawing, Complex z){
        double d;
        double d1;
        double d2;
        if( z.isFinite() ){
            double d3 = sqr( abs(z) );
            double d5 = 1.0D + d3;
            d = Re(z) / d5;
            d1 = d3 / d5 - 0.5D;
            d2 = Im(z) / d5;
        } else{
            d = 0.0D;
            d1 = 0.5D;
            d2 = 0.0D;
        }
        double d4 = R.data[2][0] * d + R.data[2][1] * d1 + R.data[2][2] * d2 + R.data[2][3];
        if(d4 <= 0.0D){
            double d6 = R.data[0][0] * d + R.data[0][1] * d1 + R.data[0][2] * d2 + R.data[0][3];
            double d7 = R.data[1][0] * d + R.data[1][1] * d1 + R.data[1][2] * d2 + R.data[1][3];
            drawing.moveTo(
                    (int)(d6 * xyscale + (double)getSize().width * 0.5D),
                    (int)(-d7 * xyscale + (double)getSize().height * 0.5D)
            );
        }
    }

    @Override public void paint(Graphics g){
        g = doubleBuffer.offScreen(g);
        xyscale = (double)Math.min(getSize().width, getSize().height) * 0.80000000000000004D;
        Drawing drawing = new Drawing(g);
        drawing.drawCircle(getSize().width / 2, getSize().height / 2, 0.5D * xyscale);
        for(int i = 0; i < marks.length; i++)
            drawComplex(drawing, marks[i]);
        w.drawStuff(drawing);
        doubleBuffer.onScreen();
    }

    Complex Point2Complex(Point point){
        double d = ((double)point.x - (double)getSize().width * 0.5D) / xyscale;
        double d1 = ((double)getSize().height * 0.5D - (double)point.y) / xyscale;
        double d2 = 0.25D - d * d - d1 * d1;
        if(d2 >= 0.0D){
            Vector3 vector3 = R1.multiply(d, d1, -Math.sqrt(d2));
            return f3dC(vector3);
        } else{
            return null;
        }
    }

    void reset(){
        R.unit();
        R1.unit();
    }

    public void setFont(Font font){
        super.setFont(font);
        int i = getFontMetrics(font).getAscent();
        MARKLENGTH = i / 5;
    }

    void shift(int i, int j){
        double d = (double)j * (6.2831853071795862D / (double)getSize().width);
        double d1 = (double)i * (6.2831853071795862D / (double)getSize().height);
        R.preRot('x', -d);
        R.preRot('y', -d1);
        R1.postRot('x', d);
        R1.postRot('y', d1);
    }

    void zoomIn(){}
    void zoomOut(){}

}

File: src/main/scala/cat/inspiracio/calculator/Sphere.scala
/*	Copyright 2011 Alexander Bunkenburg alex@inspiracio.cat
 *
 * This file is part of Complex Calculator.
 *
 * Complex Calculator is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Complex Calculator is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Complex Calculator. If not, see <http://www.gnu.org/licenses/>.
 * */
package cat.inspiracio.calculator

import java.awt.{Dimension, Font, Graphics, Point}

import cat.inspiracio.complex._
import cat.inspiracio.geometry.{Matrix44, Point2, Vector3}

// Referenced classes of package bunkenba.calculator:
//            WorldRepresentation, DoubleBuffer, Drawing, Matrix44,
//            Vector3, World

/** A world of complex numbers on the Riemann sphere. */
final class Sphere private[calculator](val world: World) extends WorldRepresentation(world) {
  import MoreGraphics.GraphicsExtended
  import Point2._

  private var markLength = 2
  private var factor: Double = 0

  //State -------------------------------------------------------

  private var R = Matrix44.One
  private var R1 = Matrix44.One

  //Methods ------------------------------------------------------

  override private[calculator] def draw(g: Graphics, c: Complex) = {
    isFrontC(c).foreach{ point =>
      g.drawCross( point, markLength )
      g.drawString(c.toString, point.x+2, point.y+2 )
    }
  }

  override private[calculator] def draw(g: Graphics, zs: List[Complex]) =
    if ( zs != null ) {
      def draw: List[Option[Point]] => Unit = {
        case Nil => ()
        case Some(a) :: Nil => ()
        case None :: xs => draw(xs)
        case Some(a) :: None :: xs => draw(xs)
        case Some(a) :: Some(b) :: xs => { g.drawLine(a, b); draw( Some(b)::xs ) }
      }
      draw(zs map isFrontC)
    }

  /** Maps 3d space to Complex */
  private def f3dC(v: Vector3): Complex =
    if (v.y == 0.5 ) ∞
    else if (v.y == -0.5 ) 0
    else Cartesian(v.x / (0.5 - v.y), v.z / (0.5 - v.y))

  /** Maps a complex number to pixel space.
    * If it is visible on the front, returns the Point, else None. */
  private def isFrontC(c: Complex): Option[Point] = {
    val v = fC3d(c)               // 3d space
    val Vector3(x, y, z) = R * v  // to 3d view space
    if ( z <= 0 )                 // front means z <= 0
      Some( f2dPoint(x, y) )      // to pixel space
    else
      None
  }

  /** Maps (x, y), the first two coordinates of 3d view space, to pixel Point. */
  private def f2dPoint(x: Double, y: Double): Point = {
    Point2(
      (x * factor + width * 0.5).toInt,
      (-y * factor + height * 0.5).toInt
    )
  }

  /** Maps Complex to 3d space */
  private def fC3d(c: Complex): (Double, Double, Double) =
    if (finite(c)) {
      val Cartesian(re, im) = c
      val a = sqr(re) + sqr(im) //sqr(abs(c))
      val d = 1 + a
      (
        re/d,
        a/d - 0.5,
        im/d
      )
    }
    else (0, 0.5, 0)

  /** Called by swing */
  override def paint(g: Graphics): Unit = {
    factor = Math.min(width, height) * 0.8
    g.drawCircle(width / 2, height / 2, 0.5 * factor)
    List[Complex](0, ∞, 1, -1, i, -i ).foreach{ draw(g, _) }
    w.draw(g) //tell the world to draw its stuff (numbers, pictlets, ...)
  }

  /** Maps a point to a complex number, if the point is on the sphere, else None. */
  override private[calculator] def point2Complex(p: Point): Option[Complex] = {
    //inverse to f2dPoint
    val x = (p.x - width * 0.5) / factor
    val y = (height * 0.5 - p.y) / factor
    val z = 0.25 - sqr(x) - sqr(y)

    if ( 0 <= z ) {
      val v = R1 * (x, y, -sqrt(z) )  // from 3d view space to 3d Riemann space
      val c = f3dC(v)                 // from 3d Riemann space to complex number
      Some(c)
    }
    else
      None
  }

  override private[calculator] def reset() = {
    R = Matrix44.One
    R1 = Matrix44.One
    repaint()
  }

  override def font_=(font: Font): Unit = {
    super.font = font
    val i = getFontMetrics(font).getAscent
    markLength = i / 5
  }

  /** The R1 matrix as it was at the start of shifting. */
  private var startR1: Matrix44 = _

  override private[calculator] def startShift(start: Point) = {
    // If the start point is not on the sphere, no shifting.
    startR1 = R1
  }

  /** This is the correct implementation. There are a lot of calculations here.
    * https://sites.google.com/site/glennmurray/Home/rotation-matrices-and-formulas/rotation-about-an-arbitrary-axis-in-3-dimensions
    * */
  override private[calculator] def shift(startPoint: Point, previousPoint: Point, endPoint: Point) = {
    import Matrix44._
    import Vector3.{angle,unit}

    if(startPoint != endPoint){
      f3dViewSpace(startPoint).foreach{ a =>
        f3dViewSpace(endPoint).foreach{ b =>

          //find rotation angle
          val theta = angle(a, b) // May be very small ... like 0.007

          //find rotation axis
          val axis = unit(a x b)
          val Vector3(u, v, w) = axis

          if (u == 0 && v == 0) {
            //special case: axis is already z-axis
            R1 = startR1 * Rz(theta)
            R = invert(R1)
          }
          else {
            val p = rotation(axis,theta)
            R1 = startR1 * p
            R = invert(R1)
          }
          repaint()
        }
      }
    }//if
  }

    /** Maps Point to 3d view space. Similar to point2Complex. */
    def f3dViewSpace(p: Point): Option[Vector3] = {
      val x = (p.x - width * 0.5) / factor
      val y = -(p.y - height * 0.5) / factor
      val z = 0.25 - sqr(x) - sqr(y)
      if ( 0 <= z ) Some(Vector3(x, y, -sqrt(z) ))  //front?
      else None
    }

}